src/OpenTelemetry/Trace/Export/BatchingActivityProcessor.cs

```diff
@@ -0,0 +1,244 @@
+            // Let's just tolerate it to avoid extra synchronization.
+            if (this.currentQueueSize >= this.maxQueueSize)
+            {
+                OpenTelemetrySdkEventSource.Log.SpanProcessorQueueIsExhausted();
+                return;
+            }
+
+            Interlocked.Increment(ref this.currentQueueSize);
+
+            this.exportQueue.Enqueue(activity);
+        }
+
+        /// <inheritdoc/>
+        public override async Task ShutdownAsync(CancellationToken cancellationToken)
+        {
+            if (!this.stopping)
+            {
+                this.stopping = true;
+
+                // This will stop the loop after current batch finishes.
+                this.cts.Cancel(false);
+                this.cts.Dispose();
+                this.cts = null;
+
+                // if there are more items, continue until cancellation token allows
+                while (this.currentQueueSize > 0 && !cancellationToken.IsCancellationRequested)
+                {
+                    await this.ExportBatchAsync(cancellationToken).ConfigureAwait(false);
+                }
+
+                await this.exporter.ShutdownAsync(cancellationToken);
+
+                // there is no point in waiting for a worker task if cancellation happens
+                // it's dead already or will die on the next iteration on its own
+
+                // ExportBatchAsync must never throw, we are here either because it was cancelled
+                // or because there are no items left
+                OpenTelemetrySdkEventSource.Log.ShutdownEvent(this.currentQueueSize);
+            }
+        }
+
+        public void Dispose()
+        {
+            this.Dispose(true);
+        }
+
+        protected virtual void Dispose(bool isDisposing)
+        {
+            if (!this.stopping)
+            {
+                this.ShutdownAsync(CancellationToken.None).ContinueWith(_ => { }).GetAwaiter().GetResult();
+            }
+
+            if (isDisposing)
+            {
+                if (this.exporter is IDisposable disposableExporter)
+                {
+                    try
+                    {
+                        disposableExporter.Dispose();
+                    }
+                    catch (Exception e)
+                    {
+                        OpenTelemetrySdkEventSource.Log.SpanProcessorException("Dispose", e);
+                    }
+                }
+            }
+        }
+
+        private async Task ExportBatchAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                List<Activity> batch = null;
+                if (this.exportQueue.TryDequeue(out var nextActivity))
+                {
+                    Interlocked.Decrement(ref this.currentQueueSize);
+                    batch = new List<Activity> { nextActivity };
```

Comment:
At first pass, it seems that it should be able to avoid allocating a new list in every run (perhaps with some extra care on shutdown). Anyway, it is an optimization we can look at it later.

Reply:
Agree and it's as it's a simple fix, I've added to my changes.